src/cdk/a11y/input-modality/input-modality-detector.spec.ts

```diff
@@ -0,0 +1,167 @@
+
+    dispatchKeyboardEvent(document, 'keydown');
+    expect(detector.mostRecentModality).toBe('keyboard');
+
+    dispatchMouseEvent(document, 'mousedown');
+    expect(detector.mostRecentModality).toBe('mouse');
+
+    dispatchTouchEvent(document, 'touchstart');
+    expect(detector.mostRecentModality).toBe('touch');
+
+    dispatchKeyboardEvent(document, 'keydown');
+    expect(detector.mostRecentModality).toBe('keyboard');
+  });
+
+  it('should emit changes in input modality', () => {
+    detector = new InputModalityDetector(platform, ngZone, document);
+    const emitted: InputModality[] = [];
+    detector.modalityChanges.subscribe((modality: InputModality) => {
+      emitted.push(modality);
+    });
+
+    expect(emitted.length).toBe(0);
+
+    dispatchKeyboardEvent(document, 'keydown');
+    expect(emitted).toEqual(['keyboard']);
+
+    dispatchKeyboardEvent(document, 'keydown');
+    expect(emitted).toEqual(['keyboard']);
+
+    dispatchMouseEvent(document, 'mousedown');
+    expect(emitted).toEqual(['keyboard', 'mouse']);
+
+    dispatchTouchEvent(document, 'touchstart');
+    expect(emitted).toEqual(['keyboard', 'mouse', 'touch']);
+
+    dispatchTouchEvent(document, 'touchstart');
+    expect(emitted).toEqual(['keyboard', 'mouse', 'touch']);
+
+    dispatchKeyboardEvent(document, 'keydown');
+    expect(emitted).toEqual(['keyboard', 'mouse', 'touch', 'keyboard']);
+  });
+
+  it('should ignore fake screen-reader mouse events', () => {
+    detector = new InputModalityDetector(platform, ngZone, document);
+
+    // Create a fake screen-reader mouse event.
+    const event = createMouseEvent('mousedown');
+    Object.defineProperty(event, 'buttons', {get: () => 0});
+    dispatchEvent(document, event);
+
+    expect(detector.mostRecentModality).toBe(null);
+  });
+
+  it('should ignore fake screen-reader touch events', () => {
+    detector = new InputModalityDetector(platform, ngZone, document);
+
+    // Create a fake screen-reader touch event.
+    const event = createTouchEvent('touchstart');
+    Object.defineProperty(event, 'touches', {get: () => [{identifier: -1}]});
+    dispatchEvent(document, event);
+
+    expect(detector.mostRecentModality).toBe(null);
+  });
+
+  it('should ignore certain modifier keys by default', () => {
+    detector = new InputModalityDetector(platform, ngZone, document);
+
+    dispatchKeyboardEvent(document, 'keydown', ALT);
+    dispatchKeyboardEvent(document, 'keydown', CONTROL);
+    dispatchKeyboardEvent(document, 'keydown', META);
+    dispatchKeyboardEvent(document, 'keydown', SHIFT);
+
+    expect(detector.mostRecentModality).toBe(null);
+  });
+
+  it('should not ignore modifier keys if specified', () => {
+    detector = new InputModalityDetector(platform, ngZone, document, {ignoreKeys: []});
+    dispatchKeyboardEvent(document, 'keydown', CONTROL);
+    expect(detector.mostRecentModality).toBe('keyboard');
+  });
+
+  it('should ignore additional keys if specified', () => {
```

Comment:
nit: maybe remove "additional"? It doesn't seem like there is a base set of keys that are ignored, if `ignoreKeys` is passed in.

Reply:
Done.